llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp

```diff
@@ -3847,15 +3847,36 @@ void DAGTypeLegalizer::ExpandIntRes_CTLZ(SDNode *N,
   Hi = DAG.getConstant(0, dl, NVT);
 }
 
-void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N,
-                                          SDValue &Lo, SDValue &Hi) {
-  SDLoc dl(N);
+void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N, SDValue &Lo, SDValue &Hi) {
+  SDValue Op = N->getOperand(0);
+  EVT VT = N->getValueType(0);
+  SDLoc DL(N);
+
+  // If the narrow CTPOP is not supported by the target, try to convert it
+  // to a libcall.
+  EVT ExpandedVT = TLI.getTypeToExpandTo(*DAG.getContext(), VT);
+  if (!TLI.isOperationLegalOrCustom(ISD::CTPOP, ExpandedVT)) {
```

Comment:
Should at least check if this is the libcall action

Reply:
I was thinking about this too, but I found [only one](https://github.com/llvm/llvm-project/blob/bb8a735facb710a07601d3bc9c45316f0de9d68d/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L1344) place where LibCall action is checked. That made me wonder if this is the right thing to do. Hopefully someone can tell for sure.

Reply:
Yes. This is a stalled migration. Ideally the Expand action would never produce a libcall.